### ISSUE TITLE
annotation metrics, what is the future direction?

### ISSUE THREAD

**Issue by kirkharris1** [kubernetes/kube-state-metrics#941]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

We make extensive use of `kube_namespace_annotations` which has been removed. This PR https://github.com/kubernetes/kube-state-metrics/pull/859 says "rethink the annotation metrics in a separate PR".   Reading around the PRs and Issues, the main concern seems to be that annotations can be numerous/large and security concerns. Would it be acceptable to reinstate annotation metrics, but with them off by default and enabled in the startup args?   Making it optional would allow users that don't have security concerns and do not have a large number of annotations to use it. 

If metrics carrying annotations will not be returning, then we will have to find another way internally so that we can upgrade beyond v1.6.0. For example, write an annotations exporter to run alongside kube-state.  Before starting on that, i thought I would check to see if there are any plans/timescales that you can share.  

Thanks

**Environment**:
- Kubernetes version (use `kubectl version`): v1.15.3
- Kube-state-metrics image version: > v1.8.0+

**Comment by lxnch** [comment]
Should this be a BUG actually? The commit that was reverted is [this](https://github.com/kubernetes/kube-state-metrics/commit/5ae00bb93fb6c224324d02d8b57c1fb1147948c9) one, which alongside the older `kube_namespace_annotations` added various other metrics for different resource (deployment, ingress etc). But the revert also removed the `kube_namespace_annotations` which is an older metric that we've ended up relying on.

**Comment by LiliC** [comment]
Thanks for this, sorry we missed to respond to this. We should find a solution for this ASAP.

 I believe allow-listing like we do with other things might be good approach. Thoughts? cc @tariq1890 @brancz

**Comment by brancz** [comment]
I think listing individual annotations that would be converted into metrics sounds appropriate.

**Comment by kirkharris1** [comment]
> I think listing individual annotations that would be converted into metrics sounds appropriate.

That would be great, it would fix the issue for our use case.

**Comment by LiliC** [comment]
@kirkharris1 with conferences coming up, we probably won't have time to work on this right now. But if you want to write up a short proposal for this and then start working on it, you are most welcome to! Do you want me to assign this issue for you? :)

**Comment by kirkharris1** [comment]
> @kirkharris1 with conferences coming up, we probably won't have time to work on this right now. But if you want to write up a short proposal for this and then start working on it, you are most welcome to! Do you want me to assign this issue for you? :)

I can't do it, but i will ask around internally.

**Comment by jw-s** [comment]
Hi @LiliC @kirkharris1, in our company we are also affected by this. I am more than happy to work on this.

**Comment by kirkharris1** [comment]
@jw-s thank you

**Comment by LiliC** [comment]
@jw-s sgtm, as agreed above we want to just allow individual annotations that do not contain any sensitive information and do not produce high unbound cardinality and convert those to metrics e.g. whitelist them in code. Let me know if you need any more info on this. Thank you!

/assign @jw-s

**Comment by jw-s** [comment]
@LiliC I suggest sticking with the previous (old) implementation of having one metric with each whitelisted/blacklisted annotation included as labels in this metric. With this proposal, we'd have to create a new cli flag specifically for this annotation unfortunately, or we go with a new metric per annotation then we can re-use the existing whitelist/blacklist cli flags.Thoughts?

**Comment by LiliC** [comment]
@jw-s Yes we would need at least one new flag to specify the resource/annotation to whitelist, we do not want to blacklist approach here.

cc @brancz @tariq1890 for thoughts

**Comment by tariq1890** [comment]
Yes. I agree with @LiliC in that we choose the approach of having a white(Allow) list. That way, users are not at the mercy of Kubernetes implementations that use annotations to store sensitive information. This provides users more control over what annotation metrics they want to allow/block as opposed to exposing all annotations and risk leaking sensitive data when they least expect it. While adding a new CLI flag may be unfortunate, it's worth making that tradeoff for security.

**Comment by lpereiraa** [comment]
Hello there, good people at Kubernetes :)
Any development on this? 
Thanks in advance!

**Comment by george-angel** [comment]
We threw this together to get by in the meantime: https://github.com/utilitywarehouse/kube-namespace-annotations-exporter

image: https://quay.io/repository/utilitywarehouse/kube-namespace-annotations-exporter

**Comment by jw-s** [comment]
> Hello there, good people at Kubernetes :)
> Any development on this?
> Thanks in advance!

I should have it finished this weekend!

**Comment by brancz** [comment]
Note we do want to reintroduce this feature but due to the various security concerns we want to make sure each annotation exposed is explicitly white listed.

**Comment by jw-s** [comment]
@brancz My understanding was we introduce a new cli flag which is a whitelist for all annotation related metrics and of course, re-introduce the annotation related metrics as discussed. Is this not the case?

**Comment by brancz** [comment]
Yes as that way the user consciously chooses the annotations to expose.

**Comment by sznmaran** [comment]
Hello, we are using old(not so reliable) version due to lack of annotation metrics. Is there any chance to implement solution described above? Should we find another way how to get annotations?
Thanks

**Comment by brancz** [comment]
There are various approaches outlined/in progress that are linked to this issue, please contribute to move this forward :)

**Comment by lilic** [comment]
My two cents on this:

Adding annotation metrics can be possible if we use the same approach as we did with labels metrics, so e.g. kube_pod_labels. So we could have the same allow mechanism as is being now implemented for those metrics, see PR https://github.com/kubernetes/kube-state-metrics/pull/1301#issue-519851559. So for annotation metrics it could be `--annotation-metric-allow-list ` flag and you pass resource name and annotation as it appears in kubernetes. By default this would be all disabled and users could just enable annotations they are interested in.

Note that this will not go into 2.0 release but we can already start the work and can be merged into master and can be included into the next release. 

@brancz @tariq1890 and anyone on this thread, would this work for you?

**Comment by brancz** [comment]
@lilic your suggestion sounds perfect. :100:

**Comment by Mepemotrarial** [comment]
What is the status of this issue? I have exactly the same use-case as @calexandre (only for cron jobs).

**Comment by sepich** [comment]
We're also affected by missing annotation metric (Labels are not enough, as they can only store [a-z0-9]).
I've took a chance to continue work of @jw-s, also addressing @lilic concerns from this thread and created new PR.
Would be glad somebody to review it, thanks.

**Comment by dgrisonnet** [comment]
This feature was implemented in https://github.com/kubernetes/kube-state-metrics/pull/1468 and is now available in kube-state-metrics v2.2.0 so we can close this issue.

/close
